cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -1449,7 +1457,13 @@ class dictionary_expression_converter : public equality_literals_collector {
       if (cudf::ast::detail::ast_operator_arity(op) == 2) {
         _dictionary_expr.push(ast::operation{op, new_operands.front(), new_operands.back()});
       } else if (cudf::ast::detail::ast_operator_arity(op) == 1) {
-        _dictionary_expr.push(ast::operation{op, new_operands.front()});
+        // If the new_operands is just a `_always_true` literal, propagate it here
+        if (auto* lit = dynamic_cast<ast::literal const*>(&new_operands.front().get());
+            lit == &_always_true) {
```

Comment:
Checking against the address of a member variable seems kind of fragile. 
This is assuming the `new_operands` do not hold copies?

Reply:
> This is assuming the new_operands do not hold copies?

Yep, the `new_operands` holds a `std::reference_wrapper<ast::expression const>`.

> Checking against the address of a member variable seems kind of fragile.

I agree but I am not really sure of another way to indicate if the child expression was evaluated (we want to apply the current op on the `new_operands`) or not (we just want to propagate a `{IDENTITY, always_true}` expr to not filter out anything). Here, `new_operands` will only dynamically cast to a literal and specifically the `_always_true` literal if the child expression was not evaluated.

Reply:
I've not come up with any alternate ideas. 
Perhaps add your comment above to the source code.
Is the dynamic cast even needed when comparing just the pointers?

Reply:
>Is the dynamic cast even needed when comparing just the pointers?

It's to check if the `new_operands` is indeed a `literal` (no need to check the pointers otherwise)